query/parser.go

```diff
@@ -213,6 +251,131 @@ func (p *Parser) addOperatorFunction() {
 	})
 }
 
+func (p *Parser) addPropertyKey(key string) {
+	p.pushNode(&evaluationContextKey{key})
+}
+
+func (p *Parser) addPropertyValue(value string) {
+	p.pushNode(&evaluationContextValue{value})
+}
+
+func (p *Parser) addEvaluationContext() {
+	p.pushNode(&evaluationContextNode{
+		api.Timerange{
+			Start:      0,
+			End:        0,
+			Resolution: 30,
+		},
+		api.SampleMean,
+		make(map[string]bool),
+	})
+}
+
+func (p *Parser) insertPropertyKeyValue() {
+	valueNode, ok := p.popNode(evaluationContextValuePointer).(*evaluationContextValue)
+	if !ok {
+		p.flagTypeAssertion()
+		return
+	}
+	keyNode, ok := p.popNode(evaluationContextKeyPointer).(*evaluationContextKey)
+	if !ok {
+		p.flagTypeAssertion()
+		return
+	}
+	contextNode, ok := p.popNode(evaluationContextNodePointer).(*evaluationContextNode)
+	if !ok {
+		p.flagTypeAssertion()
+		return
+	}
+
+	key := keyNode.key
+	value := valueNode.value
+	// Authenticate the validity of the given key and value...
+	// The key must be one of "sample"(by), "from", "to", "resolution"
+
+	// First check that the key has been assigned only once:
+	if contextNode.assigned[key] {
+		p.flagSyntaxError(SyntaxError{
+			token:   key,
+			message: fmt.Sprintf("Key %s has already been assigned", key),
+		})
+	}
+	contextNode.assigned[key] = true
+
+	switch key {
+	case "sample":
+		// If the key is "sample", it means we're in a "sample by" declaration.
+		// Only three possible sample methods are defined: min, max, or mean.
+		switch value {
+		case "'max'":
+			contextNode.SampleMethod = api.SampleMax
+		case "'min'":
+			contextNode.SampleMethod = api.SampleMin
+		case "'mean'":
+			contextNode.SampleMethod = api.SampleMean
+		default:
+			p.flagSyntaxError(SyntaxError{
+				token:   value,
+				message: fmt.Sprintf("Expected sampling method 'max', 'min', or 'mean'", value),
+			})
+		}
+	case "from":
+
```

Comment:
whitespace. and i think we can combine `from` and `to` case into one, since they do the same thing?